syntax = "proto2";
import "tensorcube.proto";

message NdArray {
    repeated uint64 shape = 1;
    repeated double values = 2;
    optional string dtype = 3;
}

message Value {
    oneof value {
        bool bool_value = 1;
        string string_value = 2;
        int64 int_value = 3;
        double float_value = 4;
        bytes bytes_value = 5;
        NdArray array_value = 6;

        // for platform-specific types with custom serialization.
        bytes binary_value = 100;
    }
}

message Method {
    optional string name = 1;
    repeated Value args = 2;
    optional string module = 3;
}

message PythonFunctions {
    optional bytes objective = 1;
    optional bytes gradient = 2;
}


message TensorcubeObjective {
    required tensorcube.Graph graph = 1;
    repeated string variables = 2;
    repeated int64 var_dims = 3;
    required string objective = 4;
    repeated string gradients = 5;
    repeated Value args = 6;
}

// Constraints

message LinearConstraint {
    optional NdArray a = 1;
    optional NdArray b = 2;
}

message FixedValueConstraint {
    optional uint64 var_index = 1;
    optional double value = 2;
}

message BoundConstraint {
    optional uint64 var_index = 1;
    optional double lower = 2;
    optional double upper = 3;
}

// Initializers

message UniformInitializer {
    optional uint64 var_dim = 1;
    optional double low = 2;
    optional double high = 3;
}

message Initializer {
    oneof initializer {
        NdArray fixed_initializer = 1;
        UniformInitializer uniform_initializer = 2;
    }
}

// Direction calculator
message SteepestDescent {
}

message ConjugateGradient {
    optional double beta = 1;
}

message Adam {
    optional double alpha = 1;
    optional double beta = 2;
}

// Steplength calculators
message QuadraticInterpolation {
}
message DecayedSteplength {
    optional int64 decay_steps = 1;
    optional double initial_steplength = 2;
}
message CircleDetection {
    optional int64 circle_length = 1;
    optional double decay_rate = 2;
}

message Steplength {
    oneof steplength_calculator {
        QuadraticInterpolation quadratic_interpolation = 1;
        DecayedSteplength decayed_steplength = 2;
        CircleDetection circle_detection = 3;
    }
}

// Stoppers. This is more general than having a set of non-exclusive data fields.
message MaxIterationsStopper {
    optional int64 max_iters = 1;
}
message InfinitesimalStepStopper {
}
message Stopper {
    oneof stopper {
        MaxIterationsStopper max_iters_stopper = 1;
        InfinitesimalStepStopper infinitesimal_step_stopper = 2;
    }
}

// Configuration
enum BinaryMappingMode {
    BINARY_MAPPING_ZERO_ONE = 1;
    BINARY_MAPPING_ONES = 2;
}

message Configuration {
    optional double epsilon = 1;
    optional BinaryMappingMode binary_mapping_mode = 2;
    optional uint64 initialization_max_iters = 3;
    optional uint64 initialization_decay_steps = 4;

    // for multi-process solvers.
    optional uint64 parallelization = 5;
}

// Diary
message MemoryDiary {
}

message StreamDiary {
    optional string stream = 1;
}
message FileDiary {
    optional string path = 1;
    optional string job_name = 2;
}

message Diary {
    oneof diary {
        MemoryDiary memory_diary = 1;
        StreamDiary stream_diary = 2;
        FileDiary file_diary = 3;
    }
}

// Got a problem.
enum ProblemType {
    PROBLEM_TYPE_FIRST_ORDER = 1;
    PROBLEM_TYPE_MIP = 2;
}

message Problem {
    optional ProblemType problem_type = 1;

    oneof objective {
        PythonFunctions python_func_objective = 10;
        Method instance_objective = 11;
        TensorcubeObjective tensorcube_objective = 12;
    }

    repeated Initializer initializers = 30;

    repeated LinearConstraint linear_inequalities = 31;
    repeated LinearConstraint linear_equalities = 32;
    repeated FixedValueConstraint fixed_value_constraints = 33;
    repeated BoundConstraint bound_constraints = 34;

    repeated BoundConstraint integer_constraints = 35;
    repeated int64 binary_constraints = 36;


    oneof direction_calculator {
        SteepestDescent steepest_descent = 50;
        ConjugateGradient conjugate_gradient = 51;
        Adam adam = 52;
    }
    repeated Steplength steplength_calculators = 70;
    repeated Stopper stoppers = 71;

    optional Configuration config = 80;
    repeated Diary diaries = 81;
}

// Used in Integer programs solvers.
message IntegerSubproblem {
    optional NdArray initializer = 1;
    repeated FixedValueConstraint fixed_value_constraints = 2;
    repeated BoundConstraint bound_constraints = 3;
    optional string diary_id = 4;
}

// Got a solution.
enum Termination {
    TERMINATION_FAILED_INITIALIZATION = 1;
    TERMINATION_MAX_ITERATION = 2;
    TERMINATION_INFINITESIMAL_STEP = 3;
    TERMINATION_SATISFIED = 4;
    TERMINATION_FAILED = 5;
}

message Solution {
    optional NdArray x = 1;
    optional Termination termination = 2;
    optional int64 iteration = 3;
    optional double objective = 4;
    optional double spent_time_ms = 5;

    map<string, Value> others = 100;
}
